src/java.base/share/classes/java/lang/runtime/Carriers.java

```diff
@@ -0,0 +1,1242 @@
+            return counts.slotCount();
+        }
+
+        /**
+         * {@return index of first long component}
+         */
+        private int longOffset() {
+            return 0;
+        }
+
+        /**
+         * {@return index of first int component}
+         */
+        private int intOffset() {
+            return longCount();
+        }
+
+        /**
+         * {@return index of first object component}
+         */
+        private int objectOffset() {
+            return longCount() + intCount();
+        }
+
+        /**
+         * {@return true if primitive components present}
+         */
+        private boolean hasPrimitives() {
+            return longCount() != 0 ||  intCount() != 0;
+        }
+    }
+
+    /**
+     * This factory class generates {@link CarrierElements} instances containing the
+     * {@link MethodHandle MethodHandles} to the constructor and accessors of a carrier
+     * object.
+     * <p>
+     * Clients can create instances by describing a carrier <em>shape</em>, that
+     * is, a {@linkplain MethodType method type} whose parameter types describe the types of
+     * the carrier component values, or by providing the parameter types directly.
+     */
+    public static class CarrierFactory {
+        /**
+         * Constructor
+         */
+        private CarrierFactory() {
+            throw new AssertionError("private constructor");
+        }
+
+        /**
+         * Cache mapping {@link MethodType} to previously defined {@link CarrierElements}.
+         */
+        private static ConcurrentHashMap<MethodType, CarrierElements>
+                methodTypeCache = new ConcurrentHashMap<>();
+
+        /**
+         * Factory method to return a {@link CarrierElements} instance that matches the shape of
+         * the supplied {@link MethodType}. The return type of the {@link MethodType} is ignored.
+         *
+         * @param methodType  {@link MethodType} whose parameter types supply the
+         *                    the shape of the carrier's components
+         *
+         * @return {@link CarrierElements} instance
+         *
+         * @throws NullPointerException is methodType is null
+         * @throws IllegalArgumentException if number of component slots exceeds maximum
+         */
+        public static CarrierElements of(MethodType methodType) {
+            Objects.requireNonNull(methodType, "methodType must be not be null");
+            MethodType constructorMT = methodType.changeReturnType(Object.class);
+            CarrierShape carrierShape = new CarrierShape(constructorMT);
+            int slotCount = carrierShape.slotCount();
+
+            if (MAX_COMPONENTS < slotCount) {
+                throw new IllegalArgumentException("Exceeds maximum number of component slots");
+            }
+
+            return methodTypeCache.computeIfAbsent(constructorMT, (mt) -> {
+                if (slotCount <= MAX_OBJECT_COMPONENTS && carrierShape.hasPrimitives()) {
```

Comment:
It would be more optimal if CarierShape was constructed inside the lambda (from `mt` parameter) and the check for slotCount done inside lambda. This way fast-path is free from these checks and lambda becomes constant lambda...